staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go

```diff
@@ -230,14 +240,49 @@ func newWatchCache(
 		versioner:           versioner,
 		groupResource:       groupResource,
 		requestBookmark:     requestBookmark,
+		stopCh:              make(chan struct{}),
 	}
 	metrics.WatchCacheCapacity.WithLabelValues(groupResource.String()).Set(float64(wc.capacity))
 	wc.cond = sync.NewCond(wc.RLocker())
 	wc.indexValidator = wc.isIndexValidLocked
 
+	wc.wg.StartWithChannel(wc.stopCh, func(stopCh <-chan struct{}) {
+		ticker := clock.NewTicker(bookmarkRequestPeriod)
+		defer ticker.Stop()
+		for {
+			// wait until bookmark is needed
+			wc.RLock()
+			for wc.resourceVersion < wc.requestedResourceVersion {
+				wc.cond.Wait()
```

Comment:
Reusing `wc.cond`  here doesn't sound desired to me here - I don't want to be woken up on every single processed event (which would effectively be the case here).

I think the pattern should be:
- only update to `requestedResourceVersion` is waking me up
- a keep being woken up until I caught up to that version

The above (together with watchcache already being quite complicated) - I strongly believe we should move that logic to a separate type (in a separate file) with clear interface, like:
```
type bookmarkRequester struct {
  RequestBookmarkIfNeeded(int64)
}
```
that doesn't really depend on anything inside the watchcache

And probably even more - already created `bookmarkRequester` should be a parameter to watchcache (and should be created in Cacher before creatng watchcache.

Reply:
make sense!